File: backend/app/schemas/pdv_operator.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

PdvOperatorRole = Literal["operator", "fiscal"]
PdvAuthorizationAction = Literal[
    "authorize_open_cash",
    "authorize_close_cash",
    "open_cash",
    "withdrawal",
    "cancel_sale",
    "discount",
]


class PdvOperatorCreate(BaseModel):
    name: str = Field(min_length=2, max_length=150)
    code: str = Field(min_length=2, max_length=30)
    pin: str = Field(min_length=4, max_length=30)
    role: PdvOperatorRole = "operator"
    can_open_cash: bool = True
    can_authorize_withdrawal: bool = False
    can_authorize_cancel: bool = False
    can_authorize_discount: bool = False
    active: bool = True
    notes: str | None = None


class PdvOperatorUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=150)
    code: str | None = Field(default=None, min_length=2, max_length=30)
    pin: str | None = Field(default=None, min_length=4, max_length=30)
    role: PdvOperatorRole | None = None
    can_open_cash: bool | None = None
    can_authorize_withdrawal: bool | None = None
    can_authorize_cancel: bool | None = None
    can_authorize_discount: bool | None = None
    active: bool | None = None
    notes: str | None = None


class PdvOperatorRead(BaseModel):
    id: int
    name: str
    code: str
    role: str
    can_open_cash: bool
    can_authorize_withdrawal: bool
    can_authorize_cancel: bool
    can_authorize_discount: bool
    active: bool
    notes: str | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PdvAuthorizationRequest(BaseModel):
    code: str = Field(min_length=1, max_length=30)
    pin: str = Field(min_length=1, max_length=30)
    action: PdvAuthorizationAction


class PdvAuthorizationResponse(BaseModel):
    authorized: bool
    operator_id: int
    operator_name: str
    role: str
    message: str

File: admin_app/admin_flutter/lib/models/pdv_operator.dart
class PdvOperator {
  const PdvOperator({
    required this.id,
    required this.name,
    required this.code,
    required this.role,
    required this.canOpenCash,
    required this.canAuthorizeWithdrawal,
    required this.canAuthorizeCancel,
    required this.canAuthorizeDiscount,
    required this.active,
    required this.createdAt,
    this.notes,
  });

  final int id;
  final String name;
  final String code;
  final String role;
  final bool canOpenCash;
  final bool canAuthorizeWithdrawal;
  final bool canAuthorizeCancel;
  final bool canAuthorizeDiscount;
  final bool active;
  final String? notes;
  final DateTime createdAt;

  bool get isFiscal => role == 'fiscal';

  factory PdvOperator.fromJson(Map<String, dynamic> json) {
    return PdvOperator(
      id: json['id'] as int,
      name: json['name'] as String,
      code: json['code'] as String,
      role: json['role'] as String,
      canOpenCash: json['can_open_cash'] as bool? ?? true,
      canAuthorizeWithdrawal:
          json['can_authorize_withdrawal'] as bool? ?? false,
      canAuthorizeCancel: json['can_authorize_cancel'] as bool? ?? false,
      canAuthorizeDiscount: json['can_authorize_discount'] as bool? ?? false,
      active: json['active'] as bool? ?? true,
      notes: json['notes'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

class PdvOperatorPayload {
  const PdvOperatorPayload({
    required this.name,
    required this.code,
    required this.role,
    required this.canOpenCash,
    required this.canAuthorizeWithdrawal,
    required this.canAuthorizeCancel,
    required this.canAuthorizeDiscount,
    required this.active,
    this.pin,
    this.notes,
  });

  final String name;
  final String code;
  final String? pin;
  final String role;
  final bool canOpenCash;
  final bool canAuthorizeWithdrawal;
  final bool canAuthorizeCancel;
  final bool canAuthorizeDiscount;
  final bool active;
  final String? notes;

  Map<String, dynamic> toCreateJson() {
    return {
      'name': name.trim(),
      'code': code.trim(),
      'pin': pin,
      'role': role,
      'can_open_cash': canOpenCash,
      'can_authorize_withdrawal': canAuthorizeWithdrawal,
      'can_authorize_cancel': canAuthorizeCancel,
      'can_authorize_discount': canAuthorizeDiscount,
      'active': active,
      'notes': notes,
    };
  }

  Map<String, dynamic> toUpdateJson() {
    return {
      'name': name.trim(),
      'code': code.trim(),
      if (pin != null && pin!.isNotEmpty) 'pin': pin,
      'role': role,
      'can_open_cash': canOpenCash,
      'can_authorize_withdrawal': canAuthorizeWithdrawal,
      'can_authorize_cancel': canAuthorizeCancel,
      'can_authorize_discount': canAuthorizeDiscount,
      'active': active,
      'notes': notes,
    };
  }
}

class PdvAuthorization {
  const PdvAuthorization({
    required this.operatorId,
    required this.operatorName,
    required this.role,
    required this.message,
  });

  final int operatorId;
  final String operatorName;
  final String role;
  final String message;

  factory PdvAuthorization.fromJson(Map<String, dynamic> json) {
    return PdvAuthorization(
      operatorId: json['operator_id'] as int,
      operatorName: json['operator_name'] as String,
      role: json['role'] as String,
      message: json['message'] as String,
    );
  }
}
